lib/Monads/trace/Trace_Monad.thy

```diff
@@ -678,6 +721,49 @@ definition whileLoopE ::
 notation (output)
   whileLoopE  ("(whileLoopE (_)//  (_))" [1000, 1000] 1000)
 
+
+section "Combinators that have conditions with side effects"
+
+definition notM :: "('s, bool) tmonad \<Rightarrow> ('s, bool) tmonad" where
+  "notM m = do c \<leftarrow> m; return (\<not> c) od"
+
+definition
+  whileM :: "('s, bool) tmonad \<Rightarrow> ('s, 'a) tmonad \<Rightarrow> ('s, unit) tmonad" where
+  "whileM C B \<equiv> do
+    c \<leftarrow> C;
+    whileLoop (\<lambda>c s. c) (\<lambda>_. do B; C od) c;
+    return ()
+  od"
```

Comment:
not sure about whether this is the PR for it, but these are not the new definition style, the wrapping is (as I was reminded recently):
```
definition name ::
  very_long_type
  where
  ...
```